File: unittests/cpp/test_tcplink_read_write.cpp
#include <gtest/gtest.h>
#include <stdlib.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>
#include "../../include/tcplink.h"

extern void exec_start();
extern void exec_finished(const char *msg);

TEST(TCPLinkReadWrite, TestReadWriteNoTimeoutSuccess)
{
    return;
    TCPLink server(20000, -1);
    TCPLink client("127.0.0.1", 20000, -1);

    int max_loops = 100;
    while (max_loops >= 0)
    {
        if (server.isReady() && client.isReady())
            break;
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
        max_loops--;
    }

    if (max_loops <= 0)
    {
        printf("Timout while trying to connect\n");
        FAIL();
    }

    int size = 1024 * 1024; // 1 MB
    uint8_t *payload = new uint8_t[size];
    for (int i = 0; i < size; i++)
    {
        payload[i] = i % 100;
    }

    printf("connected, sending data now\n");
    ASSERT_TRUE(client.write(payload, size, 123.45));

    while (!server.hasData())
    {
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }

    printf("receiving the data (server)\n");
    auto [data, timestamp] = server.readMessage();

    printf("received %ld bytes\n", data.size());
    for (int i = 0; i < size; i++)
    {
        if (data[i] != payload[i])
            FAIL();
    }
    ASSERT_FLOAT_EQ(timestamp, 123.45);
}

TEST(TCPLinkReadWrite, TestReadWriteTimeoutSuccess)
{
    
    TCPLink server(20011, 50);
    TCPLink client("127.0.0.1", 20011, 50);

    int max_loops = 1000;
    while (max_loops >= 0)
    {
        if (server.isReady() && client.isReady())
            break;
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
        max_loops--;
    }

    if (max_loops <= 0)
    {
        printf("Timout while trying to connect\n");
        FAIL();
    }

    int size = 1024 * 1024; // 1 MB
    uint8_t *payload = new uint8_t[size];
    for (int i = 0; i < size; i++)
    {
        payload[i] = i % 100;
    }

    // printf("connected, sending data now\n");
    exec_start();
    ASSERT_TRUE(client.write(payload, size, 123.45));
    exec_finished("1MB data send client->server");

    // printf("receiving the data (server)\n");
    while (!server.hasData())
    {
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
    exec_start();
    auto [data, timestamp] = server.readMessage();
    exec_finished("1MB data recv client->server");

    ASSERT_EQ(data.size(), size);
    ASSERT_FLOAT_EQ(timestamp, 123.45);
    for (int i = 0; i < size; i++)
    {
        if (data[i] != payload[i])
            FAIL();
    }
}

TEST(TCPLinkReadWrite, TestReadWriteSequence)
{
    
    TCPLink server(20011, 200);
    TCPLink client("127.0.0.1", 20011, 200);

    int max_loops = 1000;
    while (max_loops >= 0)
    {
        if (server.isReady() && client.isReady())
            break;
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
        max_loops--;
    }

    if (max_loops <= 0)
    {
        printf("Timout while trying to connect\n");
        FAIL();
    }

    int size = 1024; // 1 MB
    uint8_t *payload = new uint8_t[size];
    for (int i = 0; i < size; i++)
    {
        payload[i] = 0;
    }

    // printf("connected, sending data now\n");
    exec_start();
    for (int i = 0; i < 100; i++)
    {
        payload[0] = i;
        ASSERT_TRUE(client.write(payload, size, 123.45));
    }
    exec_finished("write sequence sent client->server");

    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    // printf("receiving the data (server)\n");
    if (!server.hasData())
    {
        FAIL();
    }
    exec_start();
    int pos = 0;
    while (server.hasData())
    {
        auto [data, timestamp] = server.readMessage();
        // exec_finished("read data");

        ASSERT_EQ(data.size(), size);
        ASSERT_FLOAT_EQ(timestamp, 123.45);
        for (int i = 1; i < size; i++)
        {
            if (data[i] != 0)
            {
                printf("data #%d failed: data[%d] (%d) != 0\n", pos, i, data[i]);
                FAIL();
            }
        }
        if (data[0] != pos)
        {
            printf("data #%d failed: data[0] (%d) != %d\n", pos, data[0], pos);
        }
        // printf ("data #%d ok\n", pos);

        pos++;
    }
}

TEST(TCPLinkReadWrite, TestReadWriteOnceReadSequence)
{
    TCPLink server(20011, 200);
    TCPLink client("127.0.0.1", 20011, 200);

    int max_loops = 1000;
    while (max_loops >= 0)
    {
        if (server.isReady() && client.isReady())
            break;
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
        max_loops--;
    }

    if (max_loops <= 0)
    {
        printf("Timout while trying to connect\n");
        FAIL();
    }

    int size = 1024; // 1 MB
    uint8_t *payload = new uint8_t[size];
    for (int i = 0; i < size; i++)
    {
        payload[i] = 0;
    }

    // printf("connected, sending data now\n");
    exec_start();
    payload[0] = 14;
    ASSERT_TRUE(client.write(payload, size, 123.45));
    exec_finished("write sequence sent client->server");

    // printf("receiving the data (server)\n");
    while (!server.hasData())
    {
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
    exec_start();
    int pos = 0;
    while (server.hasData())
    {
        auto [data, timestamp] = server.readMessage();
        // exec_finished("read data");

        ASSERT_EQ(data.size(), size);
        ASSERT_FLOAT_EQ(timestamp, 123.45);
        for (int i = 1; i < size; i++)
        {
            if (data[i] != 0)
            {
                printf("data #%d failed: data[%d] (%d) != 0\n", pos, i, data[i]);
                FAIL();
            }
        }
        if (data[0] != 14)
        {
            printf("data #%d failed: data[0] (%d) != %d\n", pos, data[0], pos);
        }
        // printf ("data #%d ok\n", pos);
        pos++;
    }

    ASSERT_EQ(pos, 1);
}

TEST(TCPLinkReadWrite, TestDoubleReadWithoutDataRepeat)
{
    std::thread read_thread([]()
                            {
                                int size = 1024 * 1024; // 1 MB
                                uint8_t *payload = new uint8_t[size];
                                for (int i = 0; i < size; i++)
                                {
                                    payload[i] = i % 100;
                                }

                                TCPLink client("127.0.0.1", 20011, -1);
                                ASSERT_TRUE(client.writeWithAck(payload, size, 123.45));
                            });

    TCPLink server(20011, -1);

    auto [data, timestamp] = server.readMessageWithAck();
    printf("message received\n");

     int rcv_size = 1024 * 1024;

    ASSERT_EQ(data.size(), rcv_size);
    ASSERT_FLOAT_EQ(timestamp, 123.45);
    for (int i = 0; i < rcv_size; i++)
    {
        if (data[i] != i % 100)
            FAIL();
    }

    ASSERT_FALSE(server.hasData());
    auto [data2, timestamp2] = server.readMessage();
    ASSERT_FLOAT_EQ(timestamp2, 0);
    ASSERT_EQ(data2.size(), 0);

        // exec_start();
    read_thread.join();
    // exec_finished("1MB data recv client->server");

}

File: unittests/python/test_tcplink_read_write.py
from pydatalink import Datalink
import unittest, time


class TestDataLinkReadWrite(unittest.TestCase):
    
    _exec_start: float

    def execute_start(self) -> None:
        self._exec_start = time.time()
    
    def execute_end(self, msg: str) -> None:
        t = 1000*(time.time() - self._exec_start)
        print(f"{msg} executed in {t:.2f} ms")


    def test_read_write_with_timeout(self):
        client = Datalink(host="127.0.0.1", port=20101, timeout=50)
        server = Datalink(port=20101, timeout=50)

        start = time.time()
        max_ms = 1000
        total_exec_time = -1

        while True:
            if client.is_ready() and server.is_ready():
                break
            total_exec_time = 1000*(time.time() - start)
            if total_exec_time > max_ms:
                self.fail("timeout connecting server and client")
                return
        
        start = time.time()
        size = 1024 * 1024
        payload = "A" * size
        self.assertTrue(client.write(payload, 123.45))

        while not server.has_data():
            time.sleep(0.01)

        data, sz, timestamp = server.read()
        total_exec_time = 1000*(time.time() - start)
        print (f"data transfer executed in {total_exec_time:.2f} ms")
        self.assertEqual(size, sz)
        self.assertAlmostEqual(timestamp, 123.45, places=2)
        for i in range(size):
            if data[i] != payload[i]:
                self.fail("data mismatch")

        print("END 2")

    def test_double_read_without_data_repeat(self):
        server = Datalink(port=20011, timeout=50)
        client = Datalink(host="127.0.0.1", port=20011, timeout=50)

        max_loops = 1000
        while max_loops >= 0:
            if server.is_ready() and client.is_ready():
                break
            time.sleep(0.001)
            max_loops -= 1

        if max_loops <= 0:
            self.fail("Timeout while trying to connect")

        size = 1024 * 1024  # 1 MB
        payload = bytes([i % 100 for i in range(size)])

        self.execute_start()
        self.assertTrue(client.write(payload, 123.45))
        self.execute_end("1MB data send client->server")

        while not server.has_data():
            time.sleep(0.001)
        
        self.execute_start()
        data, timestamp = server.recv_object()
        self.execute_end("1MB data recv client->server")

        self.assertEqual(sz, size)
        self.assertAlmostEqual(timestamp, 123.45, places=2)
        self.assertEqual(data, payload)

        self.assertFalse(server.has_data())
        data2 = server.recv_object()
        self.assertEqual(timestamp2, 0.0)
        self.assertEqual(sz2, 0)


if __name__ == "__main__":
    unittest.main()